### ISSUE TITLE
The persistent cache optoin ("-P") seems to assume there will only every be a single instance

### ISSUE THREAD

**Issue by Rakasa** [rdesktop/rdesktop#251]
The persistent cache optoin ("-P") seems to assume there will only every be a single instance.

When opening multiple instances to different servers, display artefacts from concurrent sessions using the persistent cache option are seen in unrelated sessions.  not using the "-P" option does seem to resolve this, but at significant performance expense when remote hosts are all on slow connections.

At present, I have not determined what factors trigger this, as it is not present 100% of the time, but very frequently present.

Remote hosts for me include
 - Windows 10
 - Windows Server 2008 R2
 - Fedorda 27 using xrdp

Current client machine for testing is Fedora 27 using rdesktop v1.8.3 (appears to be the latest available from the Fedora 27 repos)

-- Rakasa

**Comment by derfian** [comment]
Thanks for the report!

I had a quick look at the RDP specs, but I could not find any hints to if these caches are supposed to be reused between all connections or just connections to a with the same user/server combination. I don't have time to look more into this problem right now, but hopefully someone else can step in.

**Comment by uglym8** [comment]
I'll handle this one.
Skimmed through sources this night and I think the solution will be trivial.
Stay tuned:)

**Comment by uglym8** [comment]
The easiest fix would be to add server IPA to the filename.

I'd like to take some time though and check the implementation to see how it comply with RDP spec as the majority of code had been written when no open RDP specification was available AFAIK.

**Comment by Rakasa** [comment]
I just wish I knew enough I could have provided a sample patch with the initial report... still learning...

-- Rakasa

**Comment by uglym8** [comment]
@Rakasa, don't worry, some day you definitely will.

Meanwhile, check, please, whether this temporary branch fixes your issue:
https://github.com/uglym8/rdesktop/commits/fix_pstcache

**Comment by Rakasa** [comment]
Working on getting the build tools on the station I use rdesktop the most extensively.

What is the server "IPA"?  My use cases often involve logging into the same server under different user IDs in parallel for load testing.

-- Rakasa

**Comment by uglym8** [comment]
Yes, it's an IP address.
I'll check sources for parallel usage (locking etc).

**Comment by Rakasa** [comment]
I just wish I knew enough I could have provided a sample patch with the initial report... still learning...

-- Rakasa

**Comment by Rakasa** [comment]
getting a crash with core...

It looks like this new build forces trying to read the password from stdin, and I never use that option (all my sessions are started via script and the script has not changed)

I'm not sure if the core would be useful, or even where to find it (new to systemd and it's hijacking)

-- Rakasa

**Comment by uglym8** [comment]
Could you just change the script and pass password explicitly via '-p' option?

**Comment by Rakasa** [comment]
I would much rather not as that puts my password both in a file on the disk and in the 'ps' output in clear text

-- Rakasa

**Comment by Rakasa** [comment]
used a bs password and ignored the failed login attempt, testing the new build

-- Rakasa

**Comment by uglym8** [comment]
>  would much rather not as that puts my password both in a file on the disk and in the 'ps' output in clear text

It won't be in cleartext in 'ps' output.
The password will be replaced with 'XXXXXX'

**Comment by Rakasa** [comment]
Seems to be working on sessions not going to the same host, however, I think I am seeing the 'not currently in use session' crash out when multiple sessions go to the same host.

-- Rakasa

**Comment by uglym8** [comment]
Glad that it works, though partially.

I'm still investigating parallel usage of persistent caches.

**Comment by Rakasa** [comment]
This is the error captured from the output when the parallel session to the same host crashes out:

X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  53 (X_CreatePixmap)
  Value in failed request:  0x0
  Serial number of failed request:  16048
  Current serial number in output stream:  16061


-- Rakasa

**Comment by uglym8** [comment]
Is this error easily reproducible?
What steps do I need to take to encounter it?
TIA.

**Comment by Rakasa** [comment]
I am able to crash it without a parallel session, application running in the remote is Kodi on Fedora 27 using xrdp/xvnc

crashes do not present with the original binary, though the artifacts from other sessions certainly do

**Comment by Rakasa** [comment]
my command lines from the scripts are:

/usr/local/bin/rdesktop -0 -g 1280x720 -r clipboard:PRIMARYCLIPBOARD -u pi -p XXXXXX -z -a 24 -T "Ravenclaw Pi" -P -r sound:local ravenclaw >/tmp/ravenclaw_rdesktop_pi.log 2>&1 &

/usr/local/bin/rdesktop -0 -g 1280x720 -r clipboard:PRIMARYCLIPBOARD -u retro -p XXXXXX -z -a 24 -T "Ravenclaw Retro" -P -r sound:local ravenclaw >/tmp/ravenclaw_rdesktop_retro.log 2>&1 &

simply working in them, switching back and forth, one of them will crash with one of those errors.  I do not know if those are the only two errors, or if more will appear in further testing.

Typical application usage in the "pi" session is Kodi, configurations and management, not trying to play any video or audio files, typical usage in the "retro" session a terminal, changing/adding filesystem mounts, installing packages, host admin stuff.  Sometimes a graphic file manager if I'm feeling lazy, but usually just the terminal window up.

-- Rakasa

**Comment by uglym8** [comment]
> I am able to crash it without a parallel session, application running in the remote is Kodi on Fedora 27 using xrdp/xvnc

> crashes do not present with the original binary, though the artifacts from other sessions certainly do

Ok.
Thank you.
I think that the crash issue should go to the separate issue as it's reproducible even without parallel session.
Please create a new issue, describe the symptoms  and we'll continue to discuss this probleb there.

Meanwhile, I'll soon finish the parallel session fix (need more time to test new locking scheme).
Stay tuned:)

**Comment by Rakasa** [comment]
How do I describe the new issue with regard to only being able to produce it in this temporary branch?

-- Rakasa

**Comment by uglym8** [comment]
It's really strange that you can get a crash only with my temporary branch as its changes are quite minimal, frankly speaking.

Could you please verify that you really don't get crashes under same workload while using master (https://github.com/rdesktop/rdesktop).

It seems that you compare the most recent build of rdesktop (master + my trivial changes) to
the rdesktop which came with your Linux distribution, which is more likely not that recent at all.

TIA.

**Comment by Rakasa** [comment]
Update:

The crash with core on start up when specifically not using the password operator I get in the master branch, looks like it is expecting/forcing the password to be specified on the command line or read from stdin regardless, and that is something I can't have in my environment.  Since it is in the master branch, I should be able to create the report for that without a problem.  Just need to nail down enough time to do so... 

The crash on just using it with certain applications I do not get with the master branch, test yesterday was again Kodi, did some library management, didn't crash, then just to give it something to process, told it to play a movie before I went to bed, this morning, connection is still present, so that crash does seem to be only in the fix_pstcache branch binary.

Doing a quick file by file compare, I found changes in 
 - rdesktop.c
 - tcp.c
 - xwin.c
 - pstcache.c

So I think more than just your pstcache changes were included in the temporary branch

-- Rakasa

**Comment by Rakasa** [comment]
Update:

The crash with core on start up when specifically not using the password operator I get in the master branch, looks like it is expecting/forcing the password to be specified on the command line or read from stdin regardless, and that is something I can't have in my environment.  Since it is in the master branch, I should be able to create the report for that without a problem.  Just need to nail down enough time to do so... 

The crash on just using it with certain applications I do not get with the master branch, test yesterday was again Kodi, did some library management, didn't crash, then just to give it something to process, told it to play a movie before I went to bed, this morning, connection is still present, so that crash does seem to be only in the fix_pstcache branch binary.

Doing a quick file by file compare, I found changes in 
 - rdesktop.c
 - tcp.c
 - xwin.c
 - pstcache.c

So I think more than just your pstcache changes were included in the temporary branch

-- Rakasa

**Comment by uglym8** [comment]
Rebased my tmp branch onto the most recent master.

Could you please be so kind to verify that you're still getting crashes.

TIA.

**Comment by Rakasa** [comment]
I'm getting access denied trying to clone the tmp branch

-- Rakasa

**Comment by uglym8** [comment]
Make sure that you've got write access to the directory you're trying to put clone into.

Just verified with
````
$ git clone https://github.com/uglym8/rdesktop.git

**Comment by Rakasa** [comment]
I was using 
```
git clone -b fix_pstcache https://github.com/rdesktop/rdesktop
```
downloading and building


-- Rakasa

**Comment by uglym8** [comment]
There's no "fix_pstcache" in https://github.com/rdesktop/rdesktop.git
Try 
````
git clone -b fix_pstcache  https://github.com/uglym8/rdesktop.git
````
instead
P.S. Pardon if you've already managed to download this branch:)

**Comment by Rakasa** [comment]
crash reproduced, three samples of the logged output look fairly consistent:
```
Clipboard(error): xclip_handle_SelectionNotify(), unable to find a textual target to satisfy RDP clipboard text request
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  53 (X_CreatePixmap)
  Value in failed request:  0x0
  Serial number of failed request:  5591
  Current serial number in output stream:  5632
```
```
Clipboard(error): xclip_handle_SelectionNotify(), unable to find a textual target to satisfy RDP clipboard text request
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  53 (X_CreatePixmap)
  Value in failed request:  0x0
  Serial number of failed request:  6219
  Current serial number in output stream:  6246
```
```
Clipboard(error): xclip_handle_SelectionNotify(), unable to find a textual target to satisfy RDP clipboard text request
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  53 (X_CreatePixmap)
  Value in failed request:  0x0
  Serial number of failed request:  9082
  Current serial number in output stream:  9128
```

Again, application was Kodi, reproduced both with using the mouse and using the arrow keys to move around.  Same actions in the master do not produce the crash, while in this build I can reproduce the crash I think in under a minute.

-- Rakasa

**Comment by uglym8** [comment]
OK.
Let me clarify again:

1) You're having a crash when not supplying "-p" (password/PIN) option in master branch?
It's really strange.
Could you please make separate ticket for this issue.
Please add exact steps to reproduce the issue.

2) Are you having the crashes only while using **Kodi on Fedora 27 using xrdp/xvnc** or in Windows as well?

2) When using "-P" (use persistent cache) option you're having crashes (at least when using Kodi on Windows host? Or still on Fedora 27?) only in my temporary branch and not in master branch.
Is it right?

3) Do you also experience crashes without supplying "-P" option in the same branch under the same workload?

4) Do you experience crashes when running several rdesktop processes in parallel connecting to the same host under different Windows user accounts?

5) Do you experience crashes even with single rdesktop process?

5) Are you connecting to Windows host directly (using name/IP) or using SSH connecting via forwarded port?

6) Could you please show the contents of your ~/.rdesktop/cache directory?
7) Could you verify that the problem persists after you remove all files from ~/.rdesktop/cache?

TIA.
Best regards.

**Comment by uglym8** [comment]
One more question: 
What's the point of running rdesktop to connect to xrdp?
Why use RDP when you can use simple RFB(VNC) (ThinLinc, tigervnc, etc) instead?
Any real advantages over VNC?

Sorry, just curious.
I haven't use xrdp before myself.

**Comment by uglym8** [comment]
I was able to reproduce this crash on Fedora 27 with xrdp (**0.9.6-1**, installed from yum repository, not the latest master).
No need to use Kodi at all.
Both my tmp branch and master crash when running with persistent cache support.
I'm going to investigate this issue further.
Stay tuned.

**Comment by Rakasa** [comment]
Very sorry for the lag in my replies, I'm trying to do this around catching up work after a power failure and basement flood.

re 1: crash with core when not using -p, yes, confirmed in master branch as well, will make the ticket for that as soon as I can.

re 2: "Kodi on Fedora 27 using xrdp/xvnc?"  that was what I found reproduced the crash the fastest, but not the only combination that reproduced it.  I've had it just typing in a terminal window.  I can't recall currently if I also got it in Windows or not, and do not see anything clear in my notes either way.  I'll update if I confirm it on Windows as well.

re 3: and 4:   have not had crashes previously without it, but that was last tried on the build that comes with Fedora/27...  link speed between sites is slow enough "-P" borders on not being optional.  I can specifically test this if you need it.

re 5: and 6: I have not done parallel testing on Windows since starting with you on the pst_cache fixes.  It was actually artefacts from a Windows session showing on screen in a Linux session gave me the initial clue to look at how the caching stores information.

re 7: connections to Windows and Linux are both over the standard RDP port (3389)

re 8: directory listing or tarball?

re 9: I can test results after wiping the directory, but have not tried yet.

re 'one more question':
 - consistency of tools used to connect
 - having the connections scripted in advanced regardless of whether the host was booted Windows or Linux (some are dual-boot hosts)
 - xrdp already handles multiple parallel sessions in its default configuration without me having to do the setup myself
 - seems faster than pure RFB, particularly with the persistent cache
 - I have not had the time to trace why xrdp/X.org config is not working and correct it (this is faster still)

re 'check the temp branch again':  will do

-- Rakasa

**Comment by uglym8** [comment]
@Rakasa,
thanks again for your very informative reply.

You don't have to verify that this issue also manifests itself on Windows as I have pinpointed the cause and this is really rdesktop's bug.

So thank you again for bringing it up:)

I asked the question N7 (Are you connecting to Windows host directly (using name/IP) or using SSH connecting via forwarded port?) just to make sure that we're getting distinct filename as it could be the same (e.g. 127.0.0.1) in the case of port forwading.

Disregard all my requests except the testing of temporary branch.

**Comment by Rakasa** [comment]
So far so good.

I get white blocks sometimes, these are more noticeable now because I am no longer getting the artefacts from other sessions, but I do recall having them in other builds rather than exclusively in this build.

I have not tested the parallel sessions to the same host, as the host I was using for that crashed with a disc failure, and I have not been able to replace the disc yet, but outside of that corner case, I believe the fix works.  Thank you so much  :)

-- Rakasa

**Comment by uglym8** [comment]
This is not a fix, it's merely a temporary hack.
I'll make a proper fix as soon as I get more spare time:)

**Comment by Rakasa** [comment]
Update, paying more attention for them, I have also noted the white blocks I described appearing in Microsoft's RDP client on Win10, so whatever it is, it is not an rdesktop exclusive issue.

I'm also beginning to think it is specific to only a server or set of servers, as now that I'm looking for it, I only see it there.

You stated you considered this (adding an IP address to the file name) a temporary hack; what do you envision as a complete/correct fix?

-- Rakasa

**Comment by uglym8** [comment]
> You stated you considered this (adding an IP address to the file name) a temporary hack; what do you envision as a complete/correct fix?
1. We'll need also to take into account a possibility that the user is connecting via forwarded connection (e.g. via SSH). In this case the host will be reported as 127.0.0.1. Which is not very helpful if you want to disctinct filenames:)
2. I may want to revise the way the cache is stored.
3. I'll have to carefully check the RDP standard complience as this part (persistent caching) was done when no RDP standards were open AFAIK. 
ATM, more or less it complies IMHO.
4. Finally, I'll have to fix (properly, I mean) the bug you encountered (ATM, it's just a hack to prove the concept).

**Comment by Rakasa** [comment]
May or may not be related to this

this is logged in output redirect:

> Core(error): tcp_recv(), SSL_read() failed with 5: Connection timed out
> Disconnected due to network error, retrying to reconnect for 70 minutes.

After this, the display is badly corrupted and no amount of moving windows around fully resolves the corrupted display, but ending the session and starting a new session does.

-- Rakasa
